accounts-db/src/io_uring/file_creator.rs

```diff
@@ -0,0 +1,550 @@
+        // to submit them for writing.
+        let ring_qsize = (buffer.as_mut().len() / write_capacity / 2).max(1) as u32;
+        let ring = IoUring::builder().build(ring_qsize)?;
+        ring.submitter()
+            .register_iowq_max_workers(&mut [MAX_IOWQ_WORKERS, 0])?;
+        Self::with_buffer_and_ring(ring, buffer, write_capacity, file_complete)
+    }
+
+    fn with_buffer_and_ring<F: FnMut(PathBuf) + 'a>(
+        ring: IoUring,
+        mut backing_buffer: B,
+        write_capacity: usize,
+        file_complete: F,
+    ) -> io::Result<Self> {
+        let buffer = backing_buffer.as_mut();
+        // Take prefix of buffer that is aligned to write_capacity
+        assert!(buffer.len() >= write_capacity);
+        let write_aligned_buf_len = buffer.len() / write_capacity * write_capacity;
+        let buffer = &mut buffer[..write_aligned_buf_len];
+
+        let buffers = IoFixedBuffer::split_buffer_chunks(buffer, write_capacity);
+        let state = FileCreatorState::new(buffers.collect(), file_complete);
+        let ring = Ring::new(ring, state);
+
+        IoFixedBuffer::register_buffer(&ring, buffer)?;
+
+        // Those are fixed file descriptor slots - OpenAt will active them by index
+        let fds = vec![-1; MAX_OPEN_FILES];
+        ring.register_files(&fds)?;
+
+        Ok(Self {
+            ring,
+            backing_buffer,
+        })
+    }
+}
+
+impl<B> FileCreator for IoUringFileCreator<'_, B> {
+    fn schedule_create_at_dir(
+        &mut self,
+        path: PathBuf,
+        mode: u32,
+        parent_dir_handle: Arc<File>,
+        contents: &mut dyn Read,
+    ) -> io::Result<()> {
+        let file_key = self.open(path, mode, Some(parent_dir_handle))?;
+        self.write_and_close(contents, file_key)
+    }
+
+    fn file_complete(&mut self, path: PathBuf) {
+        (self.ring.context_mut().file_complete)(path)
+    }
+
+    fn drain(&mut self) -> io::Result<()> {
+        let res = self.ring.drain();
+        self.ring.context().log_stats();
+        res
+    }
+}
+
+impl<B> IoUringFileCreator<'_, B> {
+    /// Schedule opening file at `path` with `mode` permissons.
+    ///
+    /// Returns key that can be used for scheduling writes for it.
+    fn open(
+        &mut self,
+        path: PathBuf,
+        mode: u32,
+        dir_handle: Option<Arc<File>>,
+    ) -> io::Result<usize> {
+        let file = PendingFile::from_path(path);
+        let path_bytes = Pin::new(file.zero_terminated_path_bytes(dir_handle.is_some()));
+
+        let file_key = self.wait_add_file(file)?;
+
+        let op = FileCreatorOp::Open(OpenOp {
+            dir_handle,
+            path_bytes,
+            mode,
+            file_key,
+        });
+        self.ring.push(op)?;
+
+        Ok(file_key)
+    }
+
+    fn wait_add_file(&mut self, file: PendingFile) -> io::Result<usize> {
+        while self.ring.context().files.len() >= self.ring.context().files.capacity() {
```

Comment:
like wait_free_buf, this should be a do/while too, else after
process_completions() you might have free space but are still going to do an
extra submit and wait

Reply:
done (hitting out of file capacity is not a commonly happening scenario, now the code will always need to call process_completions before realizing it has enough capacity... - not a problem though, I see that checking completion queue is just a few comparisons).